pennylane/decomposition/symbolic_decomposition.py

```diff
@@ -0,0 +1,202 @@
+# Copyright 2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This module contains special logic of decomposing symbolic operations."""
+
+from __future__ import annotations
+
+import functools
+
+import pennylane as qml
+
+from .controlled_decomposition import base_to_custom_ctrl_op
+from .decomposition_rule import DecompositionRule, register_resources
+from .resources import adjoint_resource_rep, resource_rep
+
+
+class AdjointDecomp(DecompositionRule):  # pylint: disable=too-few-public-methods
+    """The adjoint version of a decomposition rule."""
+
+    def __init__(self, base_decomposition: DecompositionRule):
+        self._base_decomposition = base_decomposition
+        super().__init__(self._get_impl(), self._get_resource_fn())
+
+    def _get_impl(self):
+        """The implementation of the adjoint of a gate."""
+
+        def _impl(*_, base, **__):
+            qml.adjoint(self._base_decomposition._impl)(  # pylint: disable=protected-access
+                *base.parameters, base.wires, **base.hyperparameters
+            )
+
+        return _impl
+
+    def _get_resource_fn(self):
+        """The resource function of the adjoint of a gate."""
+
+        def _resource_fn(base_class, base_params):  # pylint: disable=unused-argument
+            base_resources = self._base_decomposition.compute_resources(**base_params)
+            return {
+                adjoint_resource_rep(decomp_op.op_type, decomp_op.params): count
+                for decomp_op, count in base_resources.gate_counts.items()
+            }
+
+        return _resource_fn
+
+
+def _same_type_adjoint_resource(base_class, base_params):
+    """Resources of the adjoint of a gate whose adjoint is an instance of its own type."""
+    # This assumes that the adjoint of the gate has the same resources as the gate itself.
+    return {resource_rep(base_class, **base_params): 1}
+
+
+@register_resources(_same_type_adjoint_resource)
+def has_adjoint_decomp(*_, base, **__):
+    """Decompose the adjoint of a gate whose adjoint is an instance of its own type."""
+    base.adjoint()
+
+
+def _adjoint_adjoint_resource(*_, base_params, **__):
+    """Resources of the adjoint of the adjoint of a gate."""
+    base_class, base_params = base_params["base_class"], base_params["base_params"]
+    return {resource_rep(base_class, **base_params): 1}
```

Comment:
This might make things clearer, or less clear :sweat_smile: 
```suggestion
    base_of_base_class, base_of_base_params = base_params["base_class"], base_params["base_params"]
    return {resource_rep(base_of_base_class, **base_of_base_params): 1}
```

Reply:
I'll add a comment instead 😅